### PR TITLE
Adds moths

### PR DESCRIPTION
**Pull request by Jared-Fogle**
image

image

image

Adds moths as a playable species because maintainers said they'd only consider species from /tg/. Pretty much nothing but sprites was ported over because CM code. Also adds a good base for any future species to be added.

Unimportant TODOs:

- [ ] Ghost sprites are a bit off.
- [ ] Examine sprite doesn't show wings.

### CODE REVIEW

File: code/modules/mob/living/carbon/human/life/handle_fire.dm [hunk 1]
```diff
@@ -7,3 +7,4 @@
 	if((1 - thermal_protection) > 0.0001)
 		bodytemperature += BODYTEMP_HEATING_MAX
 		adjustFireLoss(5)
+	species?.handle_fire(src)
```
Comment (F-OS) [inline code/modules/mob/living/carbon/human/life/handle_fire.dm]: stray question mark?

File: code/modules/mob/living/carbon/human/species.dm [hunk 2]
```diff
@@ -154,6 +157,10 @@
 	H.visible_message("<span class='notice'>[H] hugs [target] to make [t_him] feel better!</span>", \
 					"<span class='notice'>You hug [target] to make [t_him] feel better!</span>", null, 4)
 
+/datum/species/proc/random_name(var/gender)
+	if(gender==FEMALE)	return capitalize(pick(first_names_female)) + " " + capitalize(pick(last_names))
```
Comment (F-OS) [inline code/modules/mob/living/carbon/human/species.dm]: SMASH THAT MFING ENTER BUTTON

File: code/modules/mob/living/carbon/human/species.dm [hunk 3]
```diff
@@ -154,6 +157,10 @@
...
+	else				return capitalize(pick(first_names_male)) + " " + capitalize(pick(last_names))
```
Comment (F-OS) [inline code/modules/mob/living/carbon/human/species.dm]: See above

File: code/modules/mob/living/carbon/human/species.dm [hunk 4]
```diff
@@ -244,7 +257,7 @@
 	brute_mod = 0.15
 	burn_mod = 1.50
 	reagent_tag = IS_HORROR
-	flags = HAS_SKIN_COLOR|NO_BREATHE|NO_POISON|HAS_LIPS|NO_PAIN|NO_SCAN|NO_POISON|NO_BLOOD|NO_SLIP|NO_CHEM_METABOLIZATION
```
Comment (F-OS) [inline code/modules/mob/living/carbon/human/species.dm]: How is this related to this PR?

File: code/modules/mob/living/carbon/human/species.dm [hunk 5]
```diff
@@ -368,7 +406,7 @@
 	poison_type = "phoron"//"oxygen"
 	insulated = 1
 
-	flags = NO_SCAN
```
Comment (F-OS) [inline code/modules/mob/living/carbon/human/species.dm]: Again, how is this related, if it's because you added a species pick system just make a special flag like IS_BLACKLISTED instead of adding whitelists for everything, seems hacky.

File: code/modules/mob/living/carbon/human/species.dm [hunk 6]
```diff
@@ -414,7 +452,7 @@
 	breath_type = "nitrogen"
 	poison_type = "oxygen"
 
-	flags = NO_SCAN|NO_BLOOD|NO_PAIN
```
Comment (F-OS) [inline code/modules/mob/living/carbon/human/species.dm]: see my comment above.

File: code/modules/mob/living/carbon/human/species.dm [hunk 7]
```diff
@@ -547,7 +585,7 @@
 	death_message = "seizes up and falls limp... But is it dead?"
 	language = "Zombie"
 	default_language = "Zombie"
-	flags = NO_PAIN|NO_BREATHE|NO_SCAN|NO_POISON
```
Comment (F-OS) [inline code/modules/mob/living/carbon/human/species.dm]: again, see my comment above.

**Comment by F-OS** [comment]
Ok all of my "How is this related" comments were on the wrong side, shit.

File: code/modules/mob/living/carbon/human/species.dm (hunk 2 above)
Comment (Jared-Fogle) [reply code/modules/mob/living/carbon/human/species.dm]: This is cut and pasted shit CM (Bay?) code. I can fix it.

File: code/modules/mob/living/carbon/human/species.dm (hunk 5 above)
Comment (F-OS) [reply code/modules/mob/living/carbon/human/species.dm]: Good point, sorry.

File: code/modules/mob/living/carbon/human/life/handle_fire.dm (hunk 1 above)
Comment (Jared-Fogle) [reply code/modules/mob/living/carbon/human/life/handle_fire.dm]: > Two new operators, ?. and ?:, have been added. These are null-conditional operators. When reading a?.b, the result will be null and will not crash the proc if a is null. When writing to a?.b, the write will simply not happen if a is null.

Comment (F-OS) [reply code/modules/mob/living/carbon/human/life/handle_fire.dm]: Wasn't implying it was bad, i just find byond interesting.

**Comment by Jared-Fogle** [comment]
@Tlaltecuhtli Did you read the files changed?

**Comment by factoryman942** [comment]
floofy bois are good

**Comment by ninjanomnom** [comment]
Time to see what sort of bad code a new CM race would entail

File: code/modules/mob/living/carbon/human/human_species.dm [hunk 8]
```diff
@@ -35,3 +35,7 @@
 
 /mob/living/carbon/human/synthetic_old/New(var/new_loc)
 	..(new_loc, "Early Synthetic")
+
+/mob/living/carbon/human/moth/New(var/new_loc)
+	moth_wings = pick(moth_wings_list - "Burnt Off")
+	..(new_loc, "Moth")
```
Comment (ninjanomnom) [inline code/modules/mob/living/carbon/human/human_species.dm]: You still have the same args as human, just do `/New(new_loc, new_species="Moth")`

File: code/modules/mob/living/carbon/human/species.dm [hunk 9]
```diff
@@ -225,6 +241,7 @@
 	brute_mod = 0.55
 	burn_mod = 0.55
 	unarmed_type = /datum/unarmed_attack/punch/strong
+	flags = IS_WHITELISTED
```
Comment (ninjanomnom) [inline code/modules/mob/living/carbon/human/species.dm]: You need to make this entirely based on config

File: code/modules/mob/living/carbon/human/update_icons.dm [hunk 10]
```diff
@@ -894,6 +916,18 @@ var/global/list/damage_icon_parts = list()
 
 			apply_overlay(TAIL_LAYER)
 
+/mob/living/carbon/human/proc/update_moth_wings()
+	remove_overlay(MOTH_WINGS_LAYER)
+	remove_underlay(MOTH_WINGS_BEHIND_LAYER)
+	
+	if(species.name == "Moth")
```
Comment (ninjanomnom) [inline code/modules/mob/living/carbon/human/update_icons.dm]: The species datum should be handling species behaviours

File: code/modules/mob/new_player/preferences_setup.dm [hunk 11]
```diff
@@ -179,7 +180,9 @@ datum/preferences
 		r_skin = red
 		g_skin = green
 		b_skin = blue
-
+	
+	proc/randomize_species_specific()
```
Comment (ninjanomnom) [inline code/modules/mob/new_player/preferences_setup.dm]: No relative pathing

**Comment by Jared-Fogle** [comment]
I'll work on all that when I can, that's a lot less than I expected.

File: code/__HELPERS/sanitize_values.dm [hunk 12]
```diff
@@ -42,6 +42,11 @@
 
 	return default
 
+/proc/sanitize_species(species, default = "Human")
+	if (species in all_species)
+		return species
+	return default
```
[the remark this answers is not in the document]
Comment (Jared-Fogle) [reply code/__HELPERS/sanitize_values.dm]: This function only exists for saving your preferences. Compile time checks would require me to change saving species from string name to the direct typepath, right?

File: code/modules/mob/living/carbon/human/human_species.dm (hunk 8 above)
Comment (Jared-Fogle) [reply code/modules/mob/living/carbon/human/human_species.dm]: If I did that, wouldn't it set the wings after spawning their character, but then not update? Would you rather I add new_species="Moth" then call update_body, or keep it like this?

**Comment by Jared-Fogle** [comment]
Ping @ninjanomnom

File: code/game/jobs/whitelist.dm [hunk 13]
```diff
@@ -28,26 +28,15 @@ var/list/whitelist = list()
 	if (!text)
 		log_misc("Failed to load config/alienwhitelist.txt")
 	else
-		alien_whitelist = text2list(text, "\n")
+		alien_whitelist["human"] = TRUE
+		alien_whitelist["Human"] = TRUE
```
[the remark this answers is not in the document]
Comment (Jared-Fogle) [reply code/game/jobs/whitelist.dm]: It's so you don't have to write "Human" in the alien whitelist. I had "human" in there because the original code had it along with "Human", and I'm worried about breaking stuff.

File: code/__HELPERS/sanitize_values.dm [hunk 14]
```diff
@@ -42,6 +42,11 @@
 
 	return default
 
+/proc/sanitize_species(species, default = "Human")
```
[the remark this answers is not in the document]
Comment (Jared-Fogle) [reply code/__HELPERS/sanitize_values.dm]: Seeing as this'd only be use in saving preferences, I'll just make it a one liner there.

File: code/__HELPERS/mobs.dm [hunk 15]
```diff
@@ -45,8 +45,8 @@ proc/random_facial_hair_style(gender, species = "Human")
 		return f_style
 
 proc/random_name(gender, species = "Human")
-	if(gender==FEMALE)	return capitalize(pick(first_names_female)) + " " + capitalize(pick(last_names))
-	else				return capitalize(pick(first_names_male)) + " " + capitalize(pick(last_names))
+	var/datum/species/S = all_species[species] || all_species["Human"]
+	return S.random_name(gender)
```
[the remark this answers is not in the document]
Comment (Jared-Fogle) [reply code/__HELPERS/mobs.dm]: I agree, but...
image

[the remark this answers is not in the document]
Comment (ninjanomnom) [reply code/__HELPERS/mobs.dm]: I did say that adding a new species would require every bit to be decent quality code. This is baby size refactor though honestly.

File: code/game/jobs/whitelist.dm (hunk 13 above)
[the remark this answers is not in the document]
Comment (ninjanomnom) [reply code/game/jobs/whitelist.dm]: Have just "Human"

File: code/game/gamemodes/cm_initialize.dm [hunk 16]
```diff
@@ -688,8 +688,10 @@ datum/game_mode/proc/initialize_special_clamps()
 		if(!istype(survivor))
 			current_survivors -= survivor
 			continue //Not a mind? How did this happen?
-
-		random_name = pick(random_name(FEMALE),random_name(MALE))
+		
+		var/mob/living/carbon/human/current = survivor.current
+		var/datum/species/species = istype(current) ? current.species : all_species[DEFAULT_SPECIES]
+		random_name = pick(species.random_name(FEMALE), species.random_name(MALE))
```
[the remark this answers is not in the document]
Comment (ninjanomnom) [reply code/game/gamemodes/cm_initialize.dm]: Oh I misread this. You are calling the proc twice though so change this to
`species.random_name(pick(FEMALE, MALE))`

File: code/modules/client/preferences.dm [hunk 17]
```diff
@@ -154,7 +157,7 @@ datum/preferences
 				if(load_character())
 					return
 	gender = pick(MALE, FEMALE)
-	real_name = random_name(gender)
+	real_name = all_species[species].random_name(gender)
```
Comment (ninjanomnom) [inline code/modules/client/preferences.dm]: I didn't mean just that one place, *all* of these need to be fixed

**Comment by Iatots** [comment]
Is this done?

**Comment by Jared-Fogle** [comment]
Yes.

**Comment by Rohesie** [comment]
Moths have been tested extensively. Code-wise they've worked fine so far.
